### ISSUE TITLE
vectorization failed on reverse copy of an array.

### ISSUE THREAD

**Issue by abique** [llvm/llvm-project#55299]
Hi,

clang-13.0.1 failed to vectorize the following code:
```C
#pragma clang loop vectorize(enable) interleave(enable)
         for (int k = 0; k < N; ++k)
            out[k] = in[N - k - 1];
```

With: `warning: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering`

I think this code is vectorizable, the compiler would need to load a vector, then reverse it using shuffle and then store it.
Eventually it will perform unaligned loads or store.

Regards,
Alex

**Comment by RKSimon** [comment]
@abique Is N a compile time constant? Do you have a godbolt link showing the entire code (the types of out / in etc.)

**Comment by abique** [comment]
I confirm that it optimize the loop correctly on apple clang 13.1.6 when targeting neon.

**Comment by abique** [comment]
It seems to work on godbolt: https://godbolt.org/z/hxEebqdcY

**Comment by abique** [comment]
I had code coverage enable maybe that could be related to that?

**Comment by abique** [comment]
Yes it seems so.
So I think that there is no auto-vectorization issue, just that code coverage makes the auto-vectorization pass fail.
Maybe the coverage pass could happen after the auto vectorization?

**Comment by abique** [comment]
Thank you very much for your time, I've reported as much information as I could, feel free to close if this is not a bug.
Regards,
Alex.

**Comment by RKSimon** [comment]
Current Codegen: https://godbolt.org/z/Tv6abcKdo - we can see the difference with/with -fcoverage-mappin

**Comment by fhahn** [comment]
Yeah, I think the store added for code coverage prevents us from generating a memory runtime check. Note that restrict on local variables are currently ignored by clang/llvm (Clang/LLVM won't `const float * __restrict in = fct2();` consider this not aliasing anything in the function).

If both pointers are passed as `__restrict` pointers to the function, no runtime checks need to be generated and Clang/LLVM will vectorize the code. (https://godbolt.org/z/er99faveG)

I think there are older related issues, which I'll try to link once I find them.
